src/java/org/apache/cassandra/service/accord/SavedCommand.java

```diff
@@ -73,536 +71,422 @@ private enum HasFields
-
-                             waitingOn,
-                             ifNotEqual(before, after, Command::writes, false),
-                             ifNotEqual(before, after, Command::durableListeners, true));
-    }
-
-    static Command reconstructFromDiff(List<LoadedDiff> diffs)
+    public static Writer<TxnId> diffWriter(Command before, Command after)
     {
-        return reconstructFromDiff(diffs, CommandSerializers.APPLIED);
+        return new DiffWriter(before, after);
     }
 
-    /**
-     * @param result is exposed because we are _not_ persisting result, since during loading or replay
-     *               we do not expect we will have to send a result to the client, and data results
-     *               can potentially contain a large number of entries, so it's best if they are not
-     *               written into the log.
-     */
-    @VisibleForTesting
-    static Command reconstructFromDiff(List<LoadedDiff> diffs, Result result)
+    public static void serialize(Command before, Command after, DataOutputPlus out, int userVersion) throws IOException
     {
-        TxnId txnId = null;
-
-        Timestamp executeAt = null;
-        SaveStatus saveStatus = null;
-        Status.Durability durability = null;
+        int flags = 0;
 
-        Ballot acceptedOrCommitted = Ballot.ZERO;
-        Ballot promised = null;
+        flags = collectFlags(before, after, Command::txnId, true, Fields.TXN_ID, flags);
+        flags = collectFlags(before, after, Command::executeAt, true, Fields.EXECUTE_AT, flags);
+        flags = collectFlags(before, after, Command::saveStatus, false, Fields.SAVE_STATUS, flags);
+        flags = collectFlags(before, after, Command::durability, false, Fields.DURABILITY, flags);
 
-        Route<?> route = null;
-        PartialTxn partialTxn = null;
-        PartialDeps partialDeps = null;
-        Seekables<?, ?> additionalKeysOrRanges = null;
+        flags = collectFlags(before, after, Command::acceptedOrCommitted, false, Fields.ACCEPTED, flags);
+        flags = collectFlags(before, after, Command::promised, false, Fields.PROMISED, flags);
 
-        WaitingOnProvider waitingOnProvider = null;
-        Writes writes = null;
-        Listeners.Immutable listeners = null;
+        flags = collectFlags(before, after, Command::route, true, Fields.ROUTE, flags);
+        flags = collectFlags(before, after, Command::partialTxn, false, Fields.PARTIAL_TXN, flags);
+        flags = collectFlags(before, after, Command::partialDeps, false, Fields.PARTIAL_DEPS, flags);
+        flags = collectFlags(before, after, Command::additionalKeysOrRanges, false, Fields.ADDITIONAL_KEYS, flags);
 
-        for (LoadedDiff diff : diffs)
+        Command.WaitingOn waitingOn = getWaitingOn(after);
+        if (waitingOn == null)
+            flags = setFieldIsNull(Fields.WAITING_ON, flags);
+        else
+            flags = setFieldChanged(Fields.WAITING_ON, flags);
```

Comment:
I don't believe this is true... if we add a new listener then we would hit this logic and say the waiting on changed, which it didn't?

Reply:
why not stay consistent and do `flags = collectFlags(before, after, SavedCommand::getWaitingOn, false, Fields.WAITING_ON, flags);`?